### PR TITLE
feat: enable hermetic library generation

### PR DESCRIPTION
**Pull request by diegomarquezp**
This is part of the effort to enable multi-version support in hand-written libraries. More context and details in [this doc](https://docs.google.com/document/d/1wrpyBtphdenM3BNelcnpBKGADYrGJUo686HXvSA0h-0/edit?pli=1&tab=t.0#heading=h.lbr9t8ls3oip).

This PR adds two workflows and their scripts:
 * `update_generation_config` which updates `generation_config.yaml` (added in this PR) to contain the latest `googleapis_committish` (source of truth of library generation) as well as other dependencies such as `gapic_generator_version`. This job runs nightly and creates a PR that modifies the config yaml.
 * `hermetic_library_generation` which reads `generation_config.yaml` and pushes a new commit to any PR whose latest commit affects `generation_config.yaml` (e.g. the nightly PR created by `update_generation_config`). It essentially translates changes in the config yaml to changes in the source code (of the generated part). This workflow will be a required check.

As an example of how they work in practice:
 1. Once per night, `update_generation_config` creates a PR that updates `generation_config.yaml` with the latest dependencies and googleapis committish
 2. `hermetic_library_generation` is triggered by this PR and generates the corresponding code changes. A new commit with the code changes is added to the PR. This step may occur for any PR affecting `generation_config.yaml`.

### Follow ups
 - [x] Add `CLOUD_JAVA_BOT_TOKEN` to repository for the hermetic generation workflow to work

### CODE REVIEW

File: .github/workflows/update_generation_config.yaml
Comment (ehsannas) [inline .github/workflows/update_generation_config.yaml:38]: typo: `update_generation_config.sh.sh`

Comment (diegomarquezp) [reply .github/workflows/update_generation_config.yaml]: Thanks for the catch! 
Fixed

**Review by ehsannas — COMMENTED** [review]
Have you run these workflows to test out?

The .sh files look good, although I'm not the best person to review bash files. Given that this is probably done consistently across other repos, would it make sense for someone else from your team to review as well?

File: .github/workflows/hermetic_library_generation.yaml
Comment (diegomarquezp) [reply .github/workflows/hermetic_library_generation.yaml]: Indeed. Corrected

**Comment by diegomarquezp** [comment]
> Have you run these workflows to test out?
> 
> The .sh files look good, although I'm not the best person to review bash files. Given that this is probably done consistently across other repos, would it make sense for someone else from your team to review as well?

@ehsannas thanks for taking a look. 

We have tested the scripts in a fork ([example PR](https://github.com/diegomarquezp/java-firestore/pull/10), [example workflow run](https://github.com/diegomarquezp/java-firestore/actions/runs/9455597678/job/26046745731)). These tests are tracked in [this internal doc](https://docs.google.com/document/d/1wrpyBtphdenM3BNelcnpBKGADYrGJUo686HXvSA0h-0/edit?pli=1&tab=t.snapwfmzmqb0).

We will also go through a review from @blakeli0 (adding now). We need both approvals before we merge this PR.

File: generation_config.yaml
Comment (blakeli0) [inline generation_config.yaml:15]: We may need a single quote for this field, see https://github.com/googleapis/google-cloud-java/blob/main/generation_config.yaml#L221

Comment (diegomarquezp) [reply generation_config.yaml]: Done

Comment (diegomarquezp) [reply generation_config.yaml]: Done

**Review by blakeli0 — APPROVED** [review]

**Review by ehsannas — APPROVED** [review]

**Pull request merged**
